### DISCUSSION TITLE
Admin Functionality

### DISCUSSION THREAD

**Discussion by JPKribs** [jellyfin/Swiftfin#1283] [Discussions]
### Summary
When I started on the Admin Dashboard, I originally set out to just do the Active Sessions, but menus and settings made for pretty quick development. So, I was able to power through a lot quickly and expand my goal a little. My new goal is to try and get as close to Web functionality as possible for settings/configuration. Since Swiftfin is an official Jellyfin app, it only makes sense **to me** for it to handle the bulk of Jellyfin's functionality. That's not an official position, just something I would personally like to have. So, who better to build it than the sole person asking for it!

I wanted to make this discussion a place to track what I am doing, any rationale I had for skipping anything, and track where I stand in the process. None of the items below have a timeline and are, frankly, not core Swiftfin functionality. As a result, I don't want to commit to any firm deadlines, but I will check everything off as I work on it and update this list.

The restriction I’m setting is that the settings and functions I want to include in Swiftfin should pertain to values that are updated frequently. For example, user management — adding, disabling, or updating passwords — is something that changes regularly. In contrast, settings like transcoding are typically configured once during the initial server setup and rarely updated afterward.

My primary goal is to give some transparency into what I am doing. My current progress can be found below.

---

- [x] **Users**
  > - See all Users | [PR](https://github.com/jellyfin/Swiftfin/pull/1287)
  > - Reset/Change User Password | [PR](https://github.com/jellyfin/Swiftfin/pull/1312)
  > - Create New User | [PR](https://github.com/jellyfin/Swiftfin/pull/1287)
  > - Delete User | [PR](https://github.com/jellyfin/Swiftfin/pull/1287)
  > - Edit User Permissions | [PR](https://github.com/jellyfin/Swiftfin/pull/1313)
  > - Edit User Media Access | [PR](https://github.com/jellyfin/Swiftfin/pull/1333)
  > - Edit User Media Deletion | [PR](https://github.com/jellyfin/Swiftfin/pull/1333)
  > - Edit User Device Access | [PR](https://github.com/jellyfin/Swiftfin/pull/1342)
  > - Edit User Live TV Settings | [PR](https://github.com/jellyfin/Swiftfin/pull/1342)
  > - Edit User Parental Ratings | [PR](https://github.com/jellyfin/Swiftfin/pull/1353)
  > - Edit User Access Schedules | [PR](https://github.com/jellyfin/Swiftfin/pull/1358)
  > - Edit User Profile | [PR](https://github.com/jellyfin/Swiftfin/pull/1328)
  > - Edit User Access Tags | [PR](https://github.com/jellyfin/Swiftfin/pull/1377)
  > - Edit User Block Tags | [PR](https://github.com/jellyfin/Swiftfin/pull/1377)
  > - Quick Connect Other Users | [PR](https://github.com/jellyfin/Swiftfin/pull/1488)

- [x] **Logs**
  > - Activity Logs View | [PR](https://github.com/jellyfin/Swiftfin/pull/1485)
  > - Server Logs View  | [PR](https://github.com/jellyfin/Swiftfin/pull/1230)

- [x] **API Keys**
  > - Create new API Key | [PR](https://github.com/jellyfin/Swiftfin/pull/1284)
  > - See All API Keys | [PR](https://github.com/jellyfin/Swiftfin/pull/1284)
  > - Copy API Keys | [PR](https://github.com/jellyfin/Swiftfin/pull/1284)
 
- [x] **Dashboard**
  > - Active Sessions | [PR](https://github.com/jellyfin/Swiftfin/pull/1230)
  > - Restart (Moved to Scheduled Tasks) | [PR](https://github.com/jellyfin/Swiftfin/pull/1230)
  > - Shutdown (Moved to Scheduled Tasks) | [PR](https://github.com/jellyfin/Swiftfin/pull/1230)
 
- [x] **Devices**
  > - See All Devices | [PR](https://github.com/jellyfin/Swiftfin/pull/1277)
  > - Delete one or more Devices | [PR](https://github.com/jellyfin/Swiftfin/pull/1277)
  
- [x] **Scheduled Tasks**
  > - See all Scheduled Tasks | [PR](https://github.com/jellyfin/Swiftfin/pull/1276)
  > - Manually Run Scheduled Tasks | [PR](https://github.com/jellyfin/Swiftfin/pull/1276)
  > - Schedule Scheduled Tasks | [PR](https://github.com/jellyfin/Swiftfin/pull/1276)
  > - Edit Existing Schedules for Scheduled Tasks | [PR](https://github.com/jellyfin/Swiftfin/pull/1276)

- [x] **Non-Setting Changes**
  > - Migrate all AdminDashboard items into their own Coordinator. | [PR](https://github.com/jellyfin/Swiftfin/pull/1300)
  > - ~HomeView Activity Indicator | [PR](https://github.com/jellyfin/Swiftfin/pull/1282)~

- [x] **Item Identification**
  > - Search Providers | [PR](https://github.com/jellyfin/Swiftfin/pull/1369)
  > - Select Provided Metadata | [PR](https://github.com/jellyfin/Swiftfin/pull/1369)

- [x] **Item Images**
  > - Search Providers | [PR](https://github.com/jellyfin/Swiftfin/pull/1345)
  > - Select Provided Image(s) | [PR](https://github.com/jellyfin/Swiftfin/pull/1345)
  > - Delete Image(s) | [PR](https://github.com/jellyfin/Swiftfin/pull/1345)
  > - Upload Image(s) | [PR](https://github.com/jellyfin/Swiftfin/pull/1345)

- [x] **Item Metadata**
  > - Edit Item Subtitles | [PR](https://github.com/jellyfin/Swiftfin/pull/1513)
  > - Refresh Metadata | [PR](https://github.com/jellyfin/Swiftfin/pull/1310)
  > - Edit Metadata | [PR](https://github.com/jellyfin/Swiftfin/pull/1323)
  > - Edit Tags | [PR](https://github.com/jellyfin/Swiftfin/pull/1336)
  > - Edit Genres | [PR](https://github.com/jellyfin/Swiftfin/pull/1336)
  > - Edit Studios | [PR](https://github.com/jellyfin/Swiftfin/pull/1336)
  > - Edit People | [PR](https://github.com/jellyfin/Swiftfin/pull/1336)

- [x] **Item Deletion**
  > - iOS / iPadOS | [PR](https://github.com/jellyfin/Swiftfin/pull/1310)
  > - tvOS | [PR](https://github.com/jellyfin/Swiftfin/pull/1348)

### Skipped

The following section are being skipped as they should be fairly static once the server has been set up. Additionally, most of these items require additional configuration at the operating system level, which would require a browser/desktop regardless.

- General Settings
- Live TV
- Networking
- Playback
- Libraries
  
---

### Completed
- https://github.com/jellyfin/Swiftfin/pull/1230
- https://github.com/jellyfin/Swiftfin/pull/1272
- https://github.com/jellyfin/Swiftfin/pull/1275
- https://github.com/jellyfin/Swiftfin/pull/1276
- https://github.com/jellyfin/Swiftfin/pull/1277
- https://github.com/jellyfin/Swiftfin/pull/1284
- https://github.com/jellyfin/Swiftfin/pull/1287
- https://github.com/jellyfin/Swiftfin/pull/1300
- https://github.com/jellyfin/Swiftfin/pull/1333
- https://github.com/jellyfin/Swiftfin/pull/1312
- https://github.com/jellyfin/Swiftfin/pull/1313
- https://github.com/jellyfin/Swiftfin/pull/1342
- https://github.com/jellyfin/Swiftfin/pull/1353
- https://github.com/jellyfin/Swiftfin/pull/1328
- https://github.com/jellyfin/Swiftfin/pull/1358
- https://github.com/jellyfin/Swiftfin/pull/1377
- https://github.com/jellyfin/Swiftfin/pull/1310
- https://github.com/jellyfin/Swiftfin/pull/1323
- https://github.com/jellyfin/Swiftfin/pull/1336
- https://github.com/jellyfin/Swiftfin/pull/1345
- https://github.com/jellyfin/Swiftfin/pull/1369
- https://github.com/jellyfin/Swiftfin/pull/1348
- https://github.com/jellyfin/Swiftfin/pull/1485
- https://github.com/jellyfin/Swiftfin/pull/1488
- https://github.com/jellyfin/Swiftfin/pull/1513

### Old Branches
- [#1278 - Split out into several smaller PRs.](https://github.com/jellyfin/Swiftfin/pull/1278)
- [#1282 - May come back later but will require a rework. This PR was very out of date.](https://github.com/jellyfin/Swiftfin/pull/1282)

**Comment by sjorge** [comment]
I was originally mostly interested in the active sessions, which has been working great by the way.

Being able to start a scan by triggering the tasks has also been useful.

Personally I wouldn't expect anything more, the config options in the web client get added/removed in sync with things being added in the server. Based on some PRs I've seen this has already been tricky sometimes.

I have a slight worry that it will bring an increased burden to try and keep in sync with the options as they get added/changed/removed in the server. (Then again I have really hard time wrapping my head around swift/objectc so maybe that perception is unfounded)

Moving things to a separate controller might be nice, these things certainly make some sense for tue iOS/iPadOS versions, but probably not for tvOS. So that might make it easier to 'exclude' there.

**Reply by JPKribs** [reply]
My **_updated_** goal for the Admin Dashboard is to focus on items you might need to view or change frequently from a mobile device — such as users, devices, and logs. These are dynamic elements that are updated regularly and may require a quick change on the go.

On the other hand, configurations related to libraries, networking, or transcoding are likely outside the scope of a mobile app like Swiftfin. These are usually set once during server setup and rarely need adjustments afterward. Moreover, many of these configurations require operating system changes, so you’d likely need to be on a computer to handle them effectively.

Fortunately, most of the heavy lifting is a one-time setup. Adding or removing dashboard buttons as needed is straightforward, and with the current View-ViewModel architecture, updating API calls looks to be very simple.

Overall, there isn’t much left to finalize on this side of the project. I’m hoping to wrap it up by the end of the year (though no guarantees). Ideally, we’ll have everything ready to include in version 1.3 alongside the much more important player manager changes.🤞

I hope that helps answer and questions/concerns you may have!

**Comment by nokaka** [comment]
Dear friends, I am very grateful to you for fixing some bugs and creating some new features in the past few months. Can you release some testflights so that we can experience it as soon as possible?

**Reply by JPKribs** [reply]
We currently have a TestFlight for the first changes to 1.3. There is no confirmed timelines for the next TestFlight release but I would predict after https://github.com/jellyfin/Swiftfin/pull/1203 is done since that is a fairly large and very important change.

**Reply by LePips** [reply]
Oh TestFlight builds aren't much, I just haven't bothered to push them. After #1336 I'll push one up.

**Reply by Gylesie** [reply]
Hi! Wouldn't you consider having a more regular TestFlight build schedule? Lately I have seen so many changes getting into main (big thanks @JPKribs) and as I see it, making them available to test through TestFlight could bring a lot of value and expose it to a bigger audience for testing in different environments.

As far as I know, the builds are already automated and could possibly deploy to TestFlight? Or am I missing something? The main idea is to make the TestFlight builds with incremental changes over one build every couple of months that may include a lot of changes, making identifying problems harder when something goes wrong.

Anyway, I am excited about all the changes, thank you all for your time!

**Reply by LePips** [reply]
At the moment, the signing certificate we use is expired. I have previously pinged to get that fixed.

**Reply by nokaka** [reply]
Maybe because this one was forgotten this  year? @anthonylavado 
![image](https://github.com/user-attachments/assets/9a1e6ca0-12fc-4445-8006-858b88047a6d)

**Reply by anthonylavado** [reply]
Nope. It has been renewed and is valid. There was a certificate cycling that I'm trying to resolve. When I created the original, I thought 5 years was enough. Clearly I was wrong as we're now at 6 as a project (and still going!).

(**edit:** this reminded me that I haven't sent in the expense for the memebership. I just paid it and didn't think anything of it)

**Reply by anthonylavado** [reply]
@LePips I'm finally back at my Mac today so we'll get this swapped out.
